### PR TITLE
Peckshield audit

### PR DESCRIPTION
**Pull request by marktoda**
Updates for the Peckshield Audit. Review commit-by-commit to see the issues fixed one-by-one

PVE-001 https://github.com/buttonwood-protocol/tranche/pull/36/commits/649c6de7384006d2cabb7fcb2b29faff0f731e1c
PVE-002 https://github.com/buttonwood-protocol/tranche/pull/36/commits/3d63f42ba6f538e6cec930a984e29c13c0f8b027
PVE-003 https://github.com/buttonwood-protocol/tranche/pull/36/commits/770cb191852a5f2654dc5f96d625f07034d5c300
PVE-004 https://github.com/buttonwood-protocol/tranche/pull/36/commits/6f5c4edd5017405a82e4765b0ff26e0d085c475f
PVE-005 https://github.com/buttonwood-protocol/tranche/pull/36/commits/56d43a60fb813ad5be75d698a5b90c4fb456d319

### CODE REVIEW

**Comment by marktoda** [comment]
[914e30ff-24b1-4548-a56d-db2f0acd350b.pdf](https://github.com/buttonwood-protocol/tranche/files/8180232/914e30ff-24b1-4548-a56d-db2f0acd350b.pdf)
Audit report for reference

File: contracts/BondController.sol [hunk 1]
```diff
@@ -268,6 +270,7 @@ contract BondController is IBondController, OwnableUpgradeable {
      */
     function getTrancheLetter(uint256 index, uint256 _trancheCount) internal pure returns (string memory) {
         bytes memory trancheLetters = bytes("ABCDEFGHIJKLMNOPQRSTUVWXY");
+        require(index < trancheLetters.length, "BondController: invalid tranche letter");
```
Comment (DRC9702) [inline contracts/BondController.sol:273]: What's the intended behavior for `index=26`?

**Comment by DRC9702** [comment]
PVE-003 and PVE-004 both link to the same commit

File: README.md [hunk 2]
```diff
@@ -109,6 +109,13 @@ compiler version is to add the following fields to your VSCode user settings:
 
 Where of course `v0.8.3+commit.8d00100c` can be replaced with any other version.
 
+## Admin Privileges
+
+Users of the Tranche contracts should note that there are some limited admin permissions. These will be controlled by a decentralized governance mechanism and include:
+
+- Ability to manually "mature" bonds, disabling deposits and distributing collateral to the tranche token holders
```
Comment (Fiddlekins) [inline README.md:116]: Manual maturation is determined by the bond creator, not the owner of BondFactory

**Review by Fiddlekins — CHANGES_REQUESTED** [review]
The readme issue I mentioned should be fixed, the other is optional

File: contracts/BondController.sol (hunk 1 above)
Comment (marktoda) [reply contracts/BondController.sol:273]: it fails

**Comment by marktoda** [comment]
> PVE-003 and PVE-004 both link to the same commit

fixed

> Why safeIncreaseAllowance instead of safeApprove?

see impl. safeApprove requires 2 calls - one to set approval to 0 and another to set approval. It suggests to just use safeIncreaseApproval if possible
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/f8bfa560e9a10ffadbc278bfd4cfe8b404e5d300/contracts/token/ERC20/utils/SafeERC20.sol#L42

File: README.md (hunk 2 above)
Comment (marktoda) [reply README.md:116]: oh. true. fixed

File: test/BondController.ts [hunk 3]
```diff
@@ -134,6 +134,28 @@ describe("Bond Controller", () => {
...
+      for (let i = 0; i < 25; i++) {
+        tranches.push(10);
+      }
+      const { bondFactory, admin, mockCollateralToken } = await loadFixture(getFixture([200, 300, 500]));
+      await expect(
+        bondFactory.connect(admin).createBond(mockCollateralToken.address, tranches, await time.secondsFromNow(10000)),
+      ).to.be.revertedWith("BondController: invalid tranche letter");
+    });
+
+    it("should fail with exactly 26 tranches", async () => {
```
Comment (Fiddlekins) [inline test/BondController.ts:148]: I think it should succeed though, refer to my other comment

File: contracts/BondController.sol (hunk 1 above)
Comment (Fiddlekins) [inline contracts/BondController.sol:273]: It checks index against trancheLetters.length, but trancheLetters excludes Z so when there are 26 tranche ratios this fails despite there being enough letters for each tranche.
People shouldn't really be messing about with 25 or 26 tranche bonds anyway, really, but from a "correctness" point of view we should correct this I think. Cheapest way gas-wise I think would be to add +1 to the condition:
`require(index < (trancheLetters.length + 1), "BondController: invalid tranche letter");`

**Review by Fiddlekins — CHANGES_REQUESTED** [review]
I'd prefer the length = 26 case to be handled "correctly" but happy to move on if you don't think it's worth it

File: contracts/BondController.sol (hunk 1 above)
Comment (marktoda) [reply contracts/BondController.sol]: it works

**Review by Fiddlekins — APPROVED** [review]

**Pull request merged**